modules/e2eping/e2eping.js

```diff
@@ -0,0 +1,334 @@
+ * participant.
+ *
+ * This class:
+ * 1. Sends periodic ping requests to all other participants in the
+ * conference.
+ * 2. Responds to ping requests from other participants.
+ * 3. Fires events with the end-to-end RTT to each participant whenever a
+ * response is received.
+ * 4. Fires analytics events with the end-to-end RTT periodically.
+ */
+export default class E2ePing {
+    /**
+     * @param {EventEmitter} eventEmitter - The object to use to emit events.
+     * @param {Function} sendMessage - The function to use to send a message.
+     * @param {Object} options
+     */
+    constructor(eventEmitter, options, sendMessage) {
+        this.eventEmitter = eventEmitter;
+        this.sendMessage = sendMessage;
+
+        // The interval at which pings will be sent (<= 0 disables sending).
+        this.pingIntervalMs = 10000;
+
+        // The interval at which analytics events will be sent.
+        this.analyticsIntervalMs = 60000;
+
+        // Maps a participant ID to its ParticipantWrapper
+        this.participants = {};
+
+        // Whether the WebRTC channel has been opened or not.
+        this.isDataChannelOpen = false;
+
+        if (options && options.e2eping) {
+            if (typeof options.e2eping.pingInterval === 'number') {
+                this.pingIntervalMs = options.e2eping.pingInterval;
+            }
+            if (typeof options.e2eping.analyticsInterval === 'number') {
+                this.analyticsIntervalMs = options.e2eping.analyticsInterval;
+            }
+
+            // We want to report at most once a ping interval.
+            if (this.analyticsIntervalMs > 0 && this.analyticsIntervalMs
+                < this.pingIntervalMs) {
+                this.analyticsIntervalMs = this.pingIntervalMs;
+            }
+        }
+        logger.info(
+            `Initializing e2e ping; pingInterval=${
+                this.pingIntervalMs}, analyticsInterval=${
+                this.analyticsIntervalMs}.`);
+    }
+
+    /**
+     * Notifies this instance that the communications channel has been opened
+     * and it can now send messages via sendMessage.
+     */
+    dataChannelOpened() {
+        this.isDataChannelOpen = true;
+
+        // We don't want to wait the whole interval before sending the first
+        // request, but we can't send it immediately after the participant joins
+        // either, because our data channel might not have initialized.
+        // So once the data channel initializes, send requests to everyone.
+        // Wait an additional 200ms to give a chance to the remote side (if it
+        // also just connected as is the case for the first 2 participants in a
+        // conference) to open its data channel.
+        for (const id in this.participants) {
+            if (this.participants.hasOwnProperty(id)) {
+                const participantWrapper = this.participants[id];
+
+                window.setTimeout(participantWrapper.sendRequest, 200);
```

Comment:
i think this is the source of your bind issues.  this is just passing the member function so the `this` won't be bound to participantWrapper, it'll be bound to where it's called from (which in this case will be `window`).  Doing it this way will fix it:

`window.setTimeout(() => participantWrapper.sendRequest(), 200);`

Reply:
I see. I think I prefer the explicit `bind` in the constructor so that you don't have to worry about it when you call the function.

Reply:
So then why bind all the functions if this one is the only being used in a way where it's invoked without `this` being properly bound?  I don't think it is common to proactively rebind `this` to member functions in case they will be passed as a standalone function.  If we're going to use the bind method, I think it should be done at the call-site when it is needed, as in:

`window.setTimeout(participantWrapper.sendRequest.bind(participantWrapper), 200);`

The issue only occurs when the function is passed as a standalone function (like as a callback here in `window.setTimeout`) and refers to `this`, so normal function calls (other places these functions are called via `participantWrapper.XXX`) work as expected without the bind.

Reply:
Generally an arrow function is the preferred way by the team, so what Brian suggested at the top would be the way to go.

Reply:
But I see this has been merged and not a big deal, so feel free to ignore my comment

Reply:
I'm with the impression that bind is the proffered way by the team. In jitsi-meet for example bind is used in most of those cases. Why do you think the arrow function is the preferred way?

Reply:
Bind was the "preferred way" before anyone in the team knew about the arrow function. If you search for usages in jitsi-meet (which is never code than lib-jitsi-meet) 90% of the cases are with the arrow function. I'm not going to argue about this as I don't care

Reply:
Hmm... Just to be clear my intention is not to start argument, my only intention is to know which is the "preferred way" in order to follow it.

Probably I'm missing something because in every React component we are doing binds. So I'm not really sure what you are talking about.

Reply:
We're doing binds in every components only to avoid binding on every single render